### ISSUE TITLE
Difficulty training model with adversarial regularization follow-up

### ISSUE THREAD

**Issue by ckap100** [inspire-group/membership-inference-evaluation#3]
Hi @lwsong,

Thank you for your suggestions. I tried them out but am still struggling to replicate the results. I have some follow-up questions that I'd like to ask you.

1) What is the adversarial regularization term that you used for your reported results in Table 2?

2) The loss used for the inference model in the published code is the MSE between predictions in the range [0, 1] and the binary feature (training data/reference data). My loss converges immediately to 0.25 and does not decrease in any further epochs. Did you also find this to be the case, or did the inference model perform better at some point during training?

3) In Algorithm 1 from the Nasr et al. paper, the classifier learns from a training set (D) and the inference model learns from the same training set (D) and a disjoint reference set from the same population (D’). Later in the paper though, the authors state that D^A (number of known members of the training set) and D’^A are used for training the inference model. Could you please help me understand how D^A and D’^A are used during training? 

![picture-1](https://user-images.githubusercontent.com/59280213/113391225-7a85d980-9393-11eb-9362-bce763812309.PNG)

4) Are the attack accuracies for a model trained with adversarial regularization reported in Table 2 (Purchase100 – 67.6% undefended, 51.6% defended and Texas100 63.0% undefended, 51.0% defended) computed from your own experiments, or are these numbers taken from the Nasr et al. paper?

5) Were you able to replicate the results of Nasr et al. for Cifar100? 

Thanks again for your time!

**Comment by lwsong** [comment]
1. For the Purchase100, I use 3.0 as the weight for adversary regularization term; for Texas100, I use 10.0 as the weight. Also, I saved all checkpoints for each training step, and finally selected the checkpoint with accuracy matching the original paper.

2. I just reran the code on Purchase100. In the first epoch, the MSE loss indeed keeps being 0.25. However, starting from the second epoch, MSE loss gradually decreases to around 0.23. More importantly, I think you should check the top 1 accuracy of the inference model. If it stays around 0.5 during all training epochs, it truly indicates that the inference attack model learns nothing during training.
![adv_reg](https://user-images.githubusercontent.com/23202219/113733004-12533280-96c8-11eb-9c30-4a402425e864.JPG)

3. From my understanding, D^A and D'^A are not used during training the model with adversarial regularization. Instead, they are used in evaluating the privacy risk of the defended model after training: the adversary is assumed to know some training samples (D^A) and some non-training samples (D'^A), it trains the inference model from the defended model's predictions over D^A and D'^A, and finally test the inference attack success over left training and test samples (excluding D^A and D'^A).

4. The values are taken from Nasr et al. paper. We also ran their attacks on our locally training models, the attack accuracies are pretty close to values reported in their paper.

5. I did not spend time training CIFAR100 models. If you meet some problems, you can try to use different weights for the adversarial regularization term. For example, if your trained model has pretty low accuracy, you should decrease the weight for the adversarial regularization. Also, as you can see in my response 1, my hyperparameter setting on Texas100 is actually different from the original paper.

**Comment by ckap100** [comment]
Thank you very much @lwsong for your thoughtful answers. I really appreciate you taking the time to reply. I’ve tried incorporating these tips but am still not able to replicate the results. Please find below my response to your comments. 

1.	For Purchase100, Nasr et al. report that the classifier has a top 1 training/testing accuracy of 92.2%/76.5% with defense and a top 1 training/testing accuracy of 100%/80.1% without defense. I haven’t been able to come close to reproducing either of these results. In fact, the top 1 training accuracy I get without defense never gets above 1%. I know something is going wrong during training, but I am not sure what the problem is. The model I am using has the same structure as the PurchaseClassifier in your file adversarial_regularization/utils.py. What were the hyperparameters that you used for training? I have tried using the hyperparameters released in Nasr et al.’s code, as well as variations such as altering the optimizer, learning rate, weight initializer, batch size, and activation functions. Additionally, I have tried training using the “random_r_purchase100.npy” saved randomization from your repository. Are these the randomizations that you used to train the models, and did you use 10% for the training data and 30% for the attack (reference) data during the original training?

2.	For Purchase100, the released code for Nasr et al. has an outer loop that conducts adversarial regularization for 20 epochs, but within each epoch the training is conducted for an additional 76 epochs. I see in your posted screenshot that you also train for 20 epochs and are showing data from the 8th epoch. Does this correspond to the 608th epoch, as in Nasr et al.’s code, or the actual 8th pass through all the training data batches? Furthermore, the Nasr et al. code trains the classifier and inference model separately for one and five warmup epochs, respectively. Did you also use the same warmup epoch amounts? I’ve tried varying these numbers, but it doesn’t help either way. Lastly, is the top 1 accuracy of each row from this screenshot the accuracy calculated for a single batch of training data within one epoch? Assuming you use 10% of the Purchase100 data for training and use a batch size of 128, then there will be 153 distinct batches to iterate through for a single epoch. This corresponds to the 153 number that I see, but since there are multiple rows for (101/153), it's unclear to me where the reported accuracies are calculated from.

3.	 Thanks for this explanation! It makes more sense now.

4.	See above.

5.	Thanks, I’ll look into this once I can replicate the Purchase100 results!

More generally, have you considered releasing your code for adversarial regularization training? I think that it would go a long way towards supporting reproducible research in this area! 

Thanks again for your time.

**Comment by lwsong** [comment]
Hi @ckap100 , I have uploaded the training code for adversarial regularization, please take a look.
1. The training code is almost the same as Nasr et al. code in their [repo](https://github.com/SPIN-UMass/ML-Privacy-Regulization), that is also why I did not include it before. The biggest issue you have right now is that your ML model on Purchase/Texas datasets learns nothing as you mention your training accuracy is never above 1%. Clearly, there should be some bugs in your code. You should fix those bugs before evaluating the privacy risks.
2. In the code, the outer loop with 20 epochs is to train the model on the whole dataset with 20 epochs. The inner loop with a range of 76 is to [train the model on every 2 batches](https://github.com/inspire-group/membership-inference-evaluation/blob/85229b9a77bfd6a6e28fd713a7aaa766c679b46e/adversarial%20regularization/training_code/purchase_defended.py#L748). I simply follow Nasr et al.'s code to print the loss and accuracy information, which is [average estimation](https://github.com/inspire-group/membership-inference-evaluation/blob/85229b9a77bfd6a6e28fd713a7aaa766c679b46e/adversarial%20regularization/training_code/purchase_defended.py#L545) over the current and previous batches.

In summary, you should first figure out why your model's training accuracy is below 1%. After fixing that bug, please go ahead to measure the privacy issues.
